src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs

```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
             compilationState.ModuleBuilderOpt.CompilationState.SetStateMachineType(method, stateMachineType);
-            return new IteratorRewriter(body, method, isEnumerable, stateMachineType, slotAllocatorOpt, compilationState, diagnostics).Rewrite();
+            var rewriter = new IteratorRewriter(body, method, isEnumerable, stateMachineType, slotAllocatorOpt, compilationState, diagnostics);
+            if (rewriter.EnsureAllSymbolsAndSignature())
+            {
+                return body;
+            }
+
+            return rewriter.Rewrite();
+        }
+
+        // returns true if any type we depend on is not present
+        protected bool EnsureAllSymbolsAndSignature()
+        {
+            bool hasErrors = false;
+
+            EnsureSpecialType(SpecialType.System_Int32, ref hasErrors);
+            EnsureSpecialMethod(SpecialMember.System_IDisposable__Dispose, ref hasErrors);
+
+            // IEnumerator<T>
+            EnsureSpecialPropertyAsMember(SpecialMember.System_Collections_Generic_IEnumerator_T__Current,
+                                            SpecialType.System_Collections_Generic_IEnumerator_T, ref hasErrors);
+
+            // IEnumerator
+            EnsureSpecialProperty(SpecialMember.System_Collections_IEnumerator__Current, ref hasErrors);
+            EnsureSpecialMethod(SpecialMember.System_Collections_IEnumerator__MoveNext, ref hasErrors);
+            EnsureSpecialMethod(SpecialMember.System_Collections_IEnumerator__Reset, ref hasErrors);
+
+            if (_isEnumerable)
+            {
+                // IEnumerable and IEnumerable<T>
+                EnsureSpecialMethod(SpecialMember.System_Collections_IEnumerable__GetEnumerator, ref hasErrors);
+                EnsureSpecialMethodAsMember(SpecialMember.System_Collections_Generic_IEnumerable_T__GetEnumerator,
+                                            SpecialType.System_Collections_Generic_IEnumerable_T, ref hasErrors);
+            }
+
+            return hasErrors;
+        }
+
+        private void EnsureSpecialProperty(SpecialMember type, ref bool hasErrors)
+        {
+            EnsureAvailableType(() => F.SpecialProperty(type).GetMethod, ref hasErrors);
+        }
+
+        private void EnsureSpecialMethod(SpecialMember type, ref bool hasErrors)
+        {
+            EnsureAvailableType(() => F.SpecialMethod(type), ref hasErrors);
+        }
+
+        private void EnsureSpecialType(SpecialType type, ref bool hasErrors)
+        {
+            EnsureAvailableType(() => F.SpecialType(type), ref hasErrors);
+        }
+
+        private void EnsureSpecialPropertyAsMember(SpecialMember property, SpecialType type, ref bool hasErrors)
+        {
+            EnsureAvailableType(() =>
+            {
+                var sType = F.SpecialType(type).Construct(_elementType);
+                return F.SpecialProperty(property).GetMethod.AsMember(sType);
```

Comment:
I don't see any good reason for calling AsMember for the purpose of checking existence of a property. Also what happens if GetMethod is null?